### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by 0x00000010**
Adds some more unit tests

To run the tests just execute `test.sh` in the root of the project

Tests expect smartpy-cli in the default location (install with `sh <(curl -s https://smartpy.io/cli/install.sh)`)

Also adds github workflow that will execute the tests on PRs, and a minimum royalties patch

### CODE REVIEW

File: smart-py/objkt_swap_v3_test.py [hunk 1]
```diff
@@ -531,3 +540,137 @@ def test_pause_contract():
...
+    # Try to Swap the OBJKT but fail because artist is not admin?
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(10000),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        valid=True,
```
Comment (0x00000010) [inline smart-py/objkt_swap_v3_test.py]: Why does this fail with:

```
Message: 'FA2_NOT_OPERATOR'
```

Is it because the v1 marketplace is the operator by default and there needs to be an intermediary step?

Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: The problem is that the v3 contract will transfer the user1 tokens to the contract address. To do that the v3 contract needs to have operator rights for that token. Check some of the v3 tests to see that i get them before swapping for swaps that do not fail

Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: See this swap in H=N. You see the it starts adding the H=N contract as operator of the token to swap and it finishes removing that right
https://tzkt.io/opLCmitnTxMZ88RxJrDq1oFQg4pj64Rgf1PDcNno53mkyEBpdSC

Comment (0x00000010) [reply smart-py/objkt_swap_v3_test.py]: I see!  Thank you!

File: smart-py/objkt_swap_v3_test.py [hunk 2]
```diff
@@ -531,3 +541,442 @@ def test_pause_contract():
...
+    # Artist 2 updates the new objkt to be managed by marketplace v3
+    scenario += objkt.update_operators(
+        [sp.variant("add_operator", objkt.operator_param.make(
+            owner=artist2.address,
+            operator=marketplaceV3.address,
+            token_id=153))]).run(sender=artist2)
+
+    # Fail to swap second objkt as second artist when royalty mismatch
+    #
+    # TODO this should fail because there is a royalty mismatch
```
Comment (0x00000010) [inline smart-py/objkt_swap_v3_test.py]: @jagracar is it possible for swap contract to check royalties on the objkt before allowing the swap?

File: smart-py/objkt_swap_v3_test.py [hunk 3]
```diff
@@ -531,3 +541,442 @@ def test_pause_contract():
...
+    #     sender = artist1.address,
+    #     valid = False
+    # )
+
+    # objkt number must not have changed
+    scenario.verify(marketplaceV1.data.objkt_id == 152)
+
+    # try to mint without metadata and fail
+    #
+    # TODO null metadata should fail but passes currently
```
Comment (0x00000010) [inline smart-py/objkt_swap_v3_test.py]: @jagracar assuming there's not really anything we can do about this?

File: smart-py/objkt_swap_v3_test.py [hunk 4]
```diff
@@ -531,3 +541,442 @@ def test_pause_contract():
...
+    ).run(
+        sender = artist1.address,
+        valid = False
+    )
+
+    # objkt number must not have changed
+    scenario.verify(marketplaceV1.data.objkt_id == 152)
+
+    # try to mint an objkt with too few royalties
+    # TODO should fail but passes currently
```
Comment (0x00000010) [inline smart-py/objkt_swap_v3_test.py]: assuming this is handled on the frontend but can still be added via bcd or taquito

File: smart-py/objkt_swap_v3_test.py [hunk 5]
```diff
@@ -531,3 +541,661 @@ def test_pause_contract():
...
+    scenario += marketplaceV3.collect(0).run(
+        sender = collector1,
+        amount = sp.mutez(1),
+        valid = False
+    )
+
+    # TODO
+    #
+    # The original functionality removed swaps from the array if the amount
+    # of items available becomes 0
```
Comment (0x00000010) [inline smart-py/objkt_swap_v3_test.py]: @jagracar this was the behaviour in v1, just want to confirm we leave the entry in the array

my personal pref is to leave it in the array with the number available set to 0 but maybe there's a reason to remove it completely

File: smart-py/objkt_swap_v3_test.py [hunk 6]
```diff
@@ -531,3 +541,661 @@ def test_pause_contract():
...
+    )
+
+    # first one still exists
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # try to collect own swap and fail
+    #
+    # TODO should not be able to collect own swap
```
Comment (0x00000010) [inline smart-py/objkt_swap_v3_test.py]: @jagracar this test should fail, it should not be possible to collect your own swap

File: smart-py/objkt_swap_v3_test.py (hunk 2 above)
Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: Nope, as discussed per DM, it's not possible to get the royalties and creator info from the v1 contract. This can only be done on the UI using the indexer or tzkt calls to get that information before submitting the swap. 
This is what is done currently. People swapping in H=N cannot enter the royalties and creator info, the UI is doing it for them.
Of course, people could use BCD to create the swap and put some different royalties and creator. That cannot be avoided right now. But the site again has a solution for that. Swaps that don't use the original royalties and creators are not displayed on the site, so people cannot collect those swaps using the site. They are invisible.

File: smart-py/objkt_swap_v3_test.py (hunk 3 above)
Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: That's all v1 contract related and we cannot touch it. We cannot create another minting contract for OBJKTs, because only the admin of the OBJKT FA2 contract can use the mint entry point, and the admin is the v1 contract.
The only control we can do it from the UI side, and i think they are already checking that.

File: smart-py/objkt_swap_v3_test.py (hunk 4 above)
Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: We could add a min_royalties parameter and put a verify in the swap entrypoint to be sure people respect that when they use BCD or other sites to create the swap.

File: smart-py/objkt_swap_v3_test.py (hunk 5 above)
Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: I discussed it already with some people. At the moment when all the swapped editions are collected, the swap editions are 0 and one cannot collect more editions. The row for that swap remain in the swaps big map.
The situation is different for cancel swaps. If one cancels a swap, the row is removed from the big map.
I find that i bit ugly. Either one just set the editions to 0 in both cases, or one removes the row in both cases.
However, after discussing it with other people, we decided to keep it as it is in v2, because changing the behaviour might have unknown consequences for indexers or tools that the community developed.

File: smart-py/objkt_swap_v3_test.py (hunk 6 above)
Comment (jagracar) [reply smart-py/objkt_swap_v3_test.py]: Ok, i will add a verify sp.sender != swap.issuer

File: smart-py/objkt_swap_v3_test.py (hunk 5 above)
Comment (0x00000010) [reply smart-py/objkt_swap_v3_test.py]: I updated the tests for this
